File: buyAndSellStockIII.py
# Time Complexity : O(N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0 
        
        
        buy1 = float('inf')
        sell1 = 0
        buy2 = float('inf')
        sell2 = 0
        
        for i in range(len(prices)):
            #best one transaction
            buy1 = min(buy1, prices[i])
            sell1 = max(sell1, prices[i] - buy1)
            #transaction of one and two
            buy2 = min(buy2, prices[i] - sell1)
            sell2 = max(sell2, prices[i]- buy2)
        
        return sell2

File: BuyAndSellStockIII.java
// Time Complexity : O(n);
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

public class BuyAndSellStockIII {
    public int maxProfit(int[] prices) {

        int buy1 = Integer.MAX_VALUE;
        int profit1 = 0;

        int buy2 = Integer.MAX_VALUE;
        int profit2 = 0;

        //System.out.println("Buy1"+" "+"Profit1"+" "+"Buy2"+" "+"Profit2");

        for (int i = 0; i < prices.length; i++) {


            buy1 = Math.min(buy1, prices[i]);
            profit1 = Math.max(profit1, prices[i] - buy1);

            buy2 = Math.min(buy2, prices[i] - profit1);
            profit2 = Math.max(profit2, prices[i] - buy2);

            // System.out.println(buy1+"        "+profit1+"     "+buy2+"     "+profit2);

        }

        return profit2;
    }
}